File: android/src/main/java/com/trustwalletcorereactnative/wallet_core/coin/Coin.kt
package com.trustwalletcorereactnative.wallet_core.coin

import com.trustwalletcorereactnative.wallet_core.error.CEError
import com.trustwalletcorereactnative.wallet_core.error.CError
import com.trustwalletcorereactnative.wallet_core.interfaces.ICoin
import com.trustwalletcorereactnative.wallet_core.util.toHex
import com.trustwalletcorereactnative.wallet_core.util.toHexByteArray
import org.json.JSONObject
import wallet.core.java.AnySigner
import wallet.core.jni.CoinType
import wallet.core.jni.HDWallet
import wallet.core.jni.PrivateKey

abstract class Coin(coinType: CoinType, derivationPath: String) : ICoin {

    var coinType: CoinType? = null
    private var derivationPath: String? = null


    companion object{
        fun createCoinByteType(coinType : Int) : Coin {
            return when(coinType){
                714 -> {
                    Binance()
                }

                0 -> {
                    Bitcoin()
                }

                1815 -> {
                    Cardano()
                }
                60 -> {
                    Ethereum()
                }
                501 -> {
                    SOL()
                }
                else -> {
                    throw CError(CEError.UnSupportedCoin.message,CEError.UnSupportedCoin.code)
                }
            }
        }
    }

    init {
        this.coinType = coinType
        this.derivationPath = derivationPath
    }

    override fun getAddress(mnemonic: String, passphrase: String): String {
        val hdWallet: HDWallet = HDWallet(mnemonic, passphrase)

        val privateKey: PrivateKey = hdWallet.getKey(coinType, derivationPath)

        return coinType!!.deriveAddress(privateKey)
    }

    override fun getSeed(mnemonic: String, passphrase: String): ByteArray {
        val hdWallet: HDWallet = HDWallet(mnemonic, passphrase)

        return hdWallet.seed()
    }

    override fun getHexPrivateKey(mnemonic: String, passphrase: String): String {
        return getRawPrivateKey(mnemonic, passphrase).toHex()
    }

    override fun getRawPrivateKey(mnemonic: String, passphrase: String): ByteArray {
        return getNativePrivateKey(mnemonic, passphrase).data()
    }

    override fun getHexPublicKey(mnemonic: String, passphrase: String): String {
        return getRawPublicKey(mnemonic, passphrase).toHex()
    }

    override fun getRawPublicKey(mnemonic: String, passphrase: String): ByteArray {
        return getNativePrivateKey(mnemonic, passphrase).getPublicKeySecp256k1(true).data()
    }

    fun getNativePrivateKey(mnemonic: String, passphrase: String): PrivateKey {
        try {
            val hdWallet: HDWallet = HDWallet(mnemonic, passphrase)

            return hdWallet.getKey(coinType, derivationPath)
        }catch (e : Exception){
            throw CError(CEError.InvalidMnemonic.code,CEError.InvalidMnemonic.message)
        }
    }

    override fun sign(
        mnemonic: String,
        passphrase: String,
        message: String
    ): String {
        val privateKey: PrivateKey = getNativePrivateKey(mnemonic, passphrase)

        return privateKey.sign(message.toHexByteArray(), coinType!!.curve()).toHex()
    }

    override fun signTransaction(
        tx: Map<String, Any>,
        mnemonic: String,
        passphrase: String,
    ): String {
        val privateKey = getHexPrivateKey(mnemonic, passphrase)

        return signTransaction(tx,privateKey)
    }

    override fun signTransaction(tx: Map<String, Any>, privateKey: String): String {
        val opJson = JSONObject(tx).toString()
        return AnySigner.signJSON(opJson, privateKey.toHexByteArray(), coinType!!.value())
    }

//    override fun multiSignTransaction(
//        mnemonic: String,
//        passphrase: String,
//        tx: Map<String, Any>,
//        privateKeys: ArrayList<ByteArray>
//    ): ArrayList<ByteArray> {
//        val signatures = arrayListOf<ByteArray>()
//
//        for (privateKey in privateKeys) {
//            val signature = signTransaction(tx,mnemonic, passphrase)
//            signatures.add(signature)
//        }
//        return signatures
//    }
}

File: ios/WalletCore/Coin/Coin.swift
//
//  Coin.swift
//  trust-wallet-core-react-native
//
//  Created by Nguyen Van Toan on 21/5/24.
//
import Foundation
import WalletCore

class Coin : ICoin {
    
    public let coinType : CoinType
    private let derivationPath : String
    
    init(coinType : CoinType , derivationPath : String){
        self.coinType = coinType
        self.derivationPath = derivationPath
    }
    
    static func createCoinByType(coinType : Int) throws -> Coin {
        switch coinType{
        case 714:
            return Binance()
        case 0:
            return BitCoin()
        case 1815:
            return Cardano()
        case 60:
            return Ethereum()
        case 501:
            return SOL()
        default:
            throw CError.UnSupportedCoin
        }
    }
    
    func getAddress(mnemonic: String, passPharse: String) throws -> String {
        let privateKey = try getNativePrivateKey(mnemonic: mnemonic, passphrase: passPharse)
        
        return coinType.deriveAddress(privateKey: privateKey)
    }
    
    func getSeed(mnemonic: String, passPharse: String) throws -> Data {
        let hdWalelt = HDWallet(mnemonic: mnemonic, passphrase: passPharse)!
        
        return hdWalelt.seed
    }
    
    func getHexPrivateKey(mnemonic: String, passPharse: String) throws -> String {
        let rawPrivateKey = try getRawPrivateKey(mnemonic: mnemonic, passPharse: passPharse)
        
        return rawPrivateKey.hexString
    }
    
    func getRawPrivateKey(mnemonic: String, passPharse: String) throws -> Data {
        let privateKey = try getNativePrivateKey(mnemonic: mnemonic, passphrase: passPharse)
        
        return privateKey.data
    }
    
    func getHexPulicKey(mnemonic: String, passPharse: String) throws -> String {
        let rawPublicKey = try getRawPublicKey(mnemonic: mnemonic, passPharse: passPharse)
        
        return rawPublicKey.hexString
    }
    
    func getRawPublicKey(mnemonic: String, passPharse: String) throws -> Data {
        let privateKey = try getNativePrivateKey(mnemonic: mnemonic, passphrase: passPharse)
        
        return privateKey.getPublicKeySecp256k1(compressed: true).data
    }
    
    func signTransaction(mnemonic: String, passPharse: String, tx: Dictionary<String, Any>) throws -> String {
        return try signTransaction(privateKey: getHexPrivateKey(mnemonic: mnemonic, passPharse: passPharse), tx: tx)
    }
    
    func getNativePrivateKey(mnemonic: String, passphrase: String) throws -> PrivateKey {
        let hdWallet: HDWallet? = HDWallet(mnemonic : mnemonic, passphrase : passphrase)
        
        guard let hdWallet =  hdWallet else {
            throw CError.InvalidMnemonic
        }

        return hdWallet.getKey(coin: coinType, derivationPath: derivationPath)
    }
    
    
    func sign(mnemonic: String, passPharse: String, message: String) throws -> String {
        guard let messageHash = Data(hexString: message) else {
            throw CError.InvalidHexArgumentError
        }
        
        let privateKey = try getNativePrivateKey(mnemonic: mnemonic, passphrase: passPharse)
        
        let sig = privateKey.sign(digest: messageHash, curve: coinType.curve)!
        
        return sig.hexString
    }
    
    func signTransaction(privateKey: String, tx : Dictionary<String, Any>) throws-> String{
        guard let rawPrivateKey = Data(hexString: privateKey) else {
            throw CError.InvalidHexArgumentError
        }
        
        guard let json = tx.toJsonString() else{
            throw CError.MissingArgumentError
        }
        
        return AnySigner.signJSON(json, key: rawPrivateKey, coin: coinType)
    }
}
